### DISCUSSION TITLE
Inter-App Table Dependencies and Migrations

### DISCUSSION THREAD

**Discussion by hr-curtissimo** [piccolo-orm/piccolo#703] [Q&A]
I'm following the advice of #420 _- What is the best way to extend BaseUser?_. I've created an app named **accounts** that has a `Profile(Table)` in it and a `ForeignKey` back to the `BaseUser`.

When I try to migrate **accounts**, I get an error because the `piccolo_user` table does not exist, even when I add the following to the `APP_CONFIG` of the **accounts** app:

```python
migration_dependencies=["piccolo.apps.user.piccolo_app"],
```

Do I misunderstand the purpose of `migration_dependencies`?

**Comment by dantownsend** [comment]
Try also adding it to `APP_REGISTRY` in `piccolo_conf.py`:

```python
APP_REGISTRY = AppRegistry(
    apps=[
        "my_app.piccolo_app",
        "piccolo.apps.user.piccolo_app",
    ]
)
```

It might be that `migration_dependencies` doesn't do anything if the app isn't in `APP_REGISTRY`.

**Reply by hr-curtissimo** [reply]
Hi, @dantownsend! Thanks for the recommendation. Unfortunately, that doesn't seem to work? Here're the files:

```python
# piccolo_conf.py
from piccolo.engine.postgres import PostgresEngine
from piccolo.conf.apps import AppRegistry


DB = PostgresEngine(config={ ... }) # All the settings, there

APP_REGISTRY = AppRegistry(
    apps=[
        "accounts.piccolo_app",
        "piccolo.apps.user.piccolo_app",
    ]
)
```

```python
# piccolo_app.py
import os

from piccolo.conf.apps import AppConfig, table_finder


CURRENT_DIRECTORY = os.path.dirname(os.path.abspath(__file__))


APP_CONFIG = AppConfig(
    app_name="accounts",
    migrations_folder_path=os.path.join(
        CURRENT_DIRECTORY, "piccolo_migrations"
    ),
    table_classes=table_finder(
        modules=["accounts.tables"],
        exclude_imported=True,
    ),
    migration_dependencies=["piccolo.apps.user.piccolo_app"],
    commands=[],
)
```

And, I get the error:

```plain
... stack trace ...
asyncpg.exceptions.UndefinedTableError: relation "piccolo_user" does not exist
```

**Reply by dantownsend** [reply]
What happens if you run `piccolo migrations forwards user`?

**Reply by hr-curtissimo** [reply]
When I run the migrations for that Piccolo app specifically, then it runs the three migrations, creating and updating the `piccolo_user` table. After that, of course, I can run the migrations for my **accounts** app. 

There's obviously a workaround for this. 😅 

I am just curious about the `migration_dependencies` setting.

**Comment by sinisaos** [comment]
@hr-curtissimo I can't produce your error and everything is fine. Maybe you forgot to run
```
piccolo migrations forwards session_auth
piccolo migrations forwards user
```
and than create user with `piccolo user create`
I uploaded the whole project created with `piccolo asgi new` and added the accounts app with `piccolo app new accounts`
[accounts.zip](https://github.com/piccolo-orm/piccolo/files/10156360/accounts.zip)

Once you download the folder, delete the migrations from the home and accounts apps, add the db credentials and run your migrations like this
```
piccolo migrations forwards session_auth
piccolo migrations forwards user

piccolo user create

piccolo migrations new accounts --auto
piccolo migrations forwards accounts

piccolo migrations new home --auto
piccolo migrations forwards home
```
and you should be fine.

**Reply by hr-curtissimo** [reply]
@sinisaos Thank you for that. My original question was about `migration_dependencies`. I know that I can explicitly run the migrations for the other Piccolo apps. I was wondering about why `migration_dependencies` didn't seem to run them for me.

(Also, I'm not using `session_auth`. :-))

**Reply by sinisaos** [reply]
Sorry my bad. I didn't understand the problem.

**Reply by hr-curtissimo** [reply]
@sinisaos Oh, no sorry needed! Thank you, so much, for your help! I'm trying out piccolo for an app rewrite that I'm doing, and am evaluating as much as I can as quickly as I can. I'm sorry if I caused you consternation!

**Reply by dantownsend** [reply]
@sinisaos Thanks for looking into it.

Let's say we have `app_2` which depends on  `app_1`. The intention of `migration_dependencies` is to run the `app_1` migrations before the `app_2` migrations.

But you've raised some interesting questions, which I haven't really thought about before.

If running `piccolo migrations forwards app_2`, does that mean it automatically runs the migrations for `app_1`? Or do you still have to specify both `piccolo migrations forwards app_1 app_2`?

Typically I just run `piccolo migrations forwards all`, so I haven't given it much thought up until now.

It's worth testing, and updating the docs accordingly.

**Reply by hr-curtissimo** [reply]
@dantownsend @sinisaos From a consumer of the library/documentation, I feel that the `migration_dependencies` would create a DAG of app migrations to run that would be topologically sorted and applied.

Another way to use `migration_dependencies` which I think would be a nice alternative is for it, again, to build that DAG, sort it, then check that all dependency migrations **have** been applied for the app that specifies the `migration_dependencies`. If they haven't all been run, then raise a migration error.

I would happily work on either of these solutions and contribute them, if y'all'd like?

**Reply by dantownsend** [reply]
I had a look at the codebase. So when using `piccolo migrations forwards all`:

https://github.com/piccolo-orm/piccolo/blob/eb6ba32e85c2357969e0131a2991b7e50a1dcfb1/piccolo/apps/migrations/commands/forwards.py#L109-L110

It does a very naive sort of the apps based on `migration_dependencies`. In other places in Piccolo, we use `graphlib` from the standard library for this kind of thing, so would be good to update this code to use it as well:

https://github.com/piccolo-orm/piccolo/blob/eb6ba32e85c2357969e0131a2991b7e50a1dcfb1/piccolo/conf/apps.py#L444-L451

If you just pass the app name to `forwards` (i.e. `piccolo forwards my_app`), then it doesn't seem to migrate the other apps in `migration_dependencies`, so we might want to change this behaviour.

**Reply by sinisaos** [reply]
That's exactly the problem. According to the docs if we use `migration_dependencies` Piccolo should run the migration for `BaseUser` first and then for the `accounts` app (in this case) , but this is not the case and then we get an error `asyncpg.exceptions.UndefinedTableError: relation "piccolo_user" does not exist`
because the migration for `BaseUser` was never started.

**Reply by hr-curtissimo** [reply]
@dantownsend @sinisaos Is it okay if I take a try at this and submit a PR?

**Reply by dantownsend** [reply]
@hr-curtissimo That would be great, thanks!

**Reply by sinisaos** [reply]
@hr-curtissimo Yes of course, thanks for your effort.

**Comment by dantownsend** [comment]
I've opened a PR (https://github.com/piccolo-orm/piccolo/pull/864) which fixes the inter-app migration sorting. Should be in the next release.
